File: mdc_lib/rp2040/arduino/mdc_lite.cpp
/**
 * @file    mdc_lite.cpp
 * @brief   Motor Driver Controller 极简调用库 —— 只管调用（send-only）实现
 *
 * 规范依据：../LITE.md（mdc_lite 极简 LITE API）
 * 复用之 mdc_lib（API.md 的打包原语），本文件只是极薄封装：函数名收窄、
 * 关注范围收窄，协议字节布局（LE / CRC8 0x07 初值 0 / [AA][CMD][LEN][DATA][CRC]）不变。
 *
 * 本文件不 include 任何 Arduino / ESP 头文件（纯 C++，C++11 兼容）。
 *
 * 四平台（esp32 / rp2040 / avr / esp8266）此文件内容完全一致。
 */

#include "mdc_lite.h"

/* ==================== 发送侧（send-only） ==================== */

uint16_t md_lite_ctrl(int32_t m0, int32_t m1, int32_t m2, int32_t m3,
                      uint8_t* out, uint16_t cap)
{
    /* 委托 mdc_lib：0x31 四通道控制帧（含 SYNC+CRC8 整帧）
     * 校验向量：ctrl(100,-200,0,300) DATA == 64 00 00 00 38 FF FF FF 00 00 00 00 2C 01 00 00 */
    return md_bin_motor_ctrl(m0, m1, m2, m3, out, cap);
}

uint16_t md_lite_stop(uint8_t* out, uint16_t cap)
{
    /* 便捷：四通道全零控制帧（急停/退出前发送，DATA=16B 全零） */
    return md_lite_ctrl(0, 0, 0, 0, out, cap);
}

uint16_t md_lite_subscribe(uint16_t interval_ms, uint8_t* out, uint16_t cap)
{
    /* 委托 mdc_lib：0x40 订阅状态上报。
     * 校验向量：subscribe(50) 帧 == AA 40 02 32 00 <crc> */
    return md_bin_subscribe(interval_ms, out, cap);
}

uint16_t md_lite_unsubscribe(uint8_t* out, uint16_t cap)
{
    /* 委托 mdc_lib：0x41 取消订阅。
     * 校验向量：unsubscribe() 帧 == AA 41 00 <crc> */
    return md_bin_unsubscribe(out, cap);
}

File: mdc_lib/esp32/micropython/mdc_lite.py
# -*- coding: utf-8 -*-
"""
mdc_lite.py — 极简调用库（只管调用 / send-only，MicroPython 零依赖版）
======================================================================

**核心思想：上位机调参、下位机执行。** 你只需要告诉下位机"发什么控制量"，
本模块把"要发送的字节"打包好交给你，由你自行 `uart.write()` 发送。
**不解析任何回包** —— 若需读回转速，请用 `mdc_lite_ctrl`（调用+回调接收）。

**只涉及 3 条命令：**
    * 0x31 MOTOR_CTRL    四通道控制目标（下位机执行）
    * 0x40 SUBSCRIBE     开启状态周期上报（收速度的前提）
    * 0x41 UNSUBSCRIBE   关闭状态上报（善后）

**依赖：** 复用同目录 `mdc_lib.py`（API.md 的打包原语），本文件只是极薄封装。
纯 MicroPython **零依赖**（无 machine / ustruct / struct / math 等任何 import），
在任意 MicroPython 板或 CPython 上均可直接 import。

**用法（串口由你实现）：**

    from machine import UART, Pin
    import mdc_lite

    uart = UART(2, baudrate=115200, tx=Pin(17), rx=Pin(16), timeout=50)
    uart.write(mdc_lite.subscribe(50))                # 先订阅（如需回读转速）
    uart.write(mdc_lite.ctrl(100, -200, 0, 300))      # 四通道目标值
    uart.write(mdc_lite.ctrl(0, 0, 0, 0))             # 归零
    uart.write(mdc_lite.unsubscribe())                # 退出前取消订阅（可选）

**各函数返回 `bytes`（完整二进制帧，含 SYNC+CRC8），直接 uart.write 即可。**
"""

# 复用同目录 mdc_lib（极简层不重复实现协议）
import mdc_lib as _md

__all__ = [
    "ctrl", "stop", "subscribe", "unsubscribe",
]

# 目标值含义随通道控制模式（协议规范 §3.3）：
#   open = PWM(±1000)  speed = RPM   pos = 0.1°(±3600=±360.0°)


def ctrl(m0, m1, m2, m3):
    """0x31 MOTOR_CTRL：四通道控制目标（int32 LE），返回完整帧 bytes。

    例：``ctrl(100, -200, 0, 300)`` 的 DATA 段 == ``64 00 00 00 38 FF FF FF
    00 00 00 00 2C 01 00 00``。等价于 mdc_lib.md_bin_motor_ctrl(...)。
    """
    return _md.md_bin_motor_ctrl(m0, m1, m2, m3)


def stop():
    """便捷：四通道全零控制帧（急停/退出前发送，0x31 DATA=16B 全零）。"""
    return ctrl(0, 0, 0, 0)


def subscribe(interval_ms):
    """0x40 SUBSCRIBE：开启状态周期上报，返回完整帧 bytes。

    :param interval_ms: 上报周期 ms，建议 >=20（固件钳位）。等价于 mdc_lib.md_bin_subscribe。
    """
    return _md.md_bin_subscribe(interval_ms)


def unsubscribe():
    """0x41 UNSUBSCRIBE：关闭状态上报，返回完整帧 bytes。"""
    return _md.md_bin_unsubscribe()
